### ISSUE TITLE
Using a custom built drone is causing the game to crash.

### ISSUE THREAD

**Issue by mrjpjr** [Regalis11/Barotrauma#1934]
I made a drone. It was working fine before, now the game crashes when I use it. Below are the crash logs and sub files.

I have also included the main sub that has the drone's control terminal, just in case you want to do in-game testing, or just in case the crash is due to a combination of both subs. This statement is based on the advice hubismal gave in response to my post on Discord, as I recently made some changes (including but not limited to inventory changes) to the main sub.

The name of the drone is RD 1.3 and the main sub with the control terminal is the Dart 2.10.

As for when the game crashed and how to possibly reproduce it, I believe it happened most often when I was trying to dock the drone.

The game version is 0.9.5.1 and I am using Windows 10.

Thanks for taking a look at it.

[mrjpjr drone sub crash files.zip](https://github.com/Regalis11/Barotrauma/files/3950386/mrjpjr.drone.sub.crash.files.zip)

- [x] I have searched the issue tracker to check if the issue has already been reported.

**Description**

**Steps To Reproduce**
If possible, describe how the developers can reproduce the issue. It is often extremely hard to fix a bug if we don't know how to make it happen.

**Version**
Which version of the game did the bug happen in. You can see the version number at the bottom left corner of the main menu, and in a file called "Version.txt" inside the game folder.
Also, please include the operating system you're using (Windows/Linux/Mac).

**Additional information**

**Comment by jamiebidelia** [comment]
Hello!  I am not a part of the Barotrauma team, but was interested in your bug and looked into it.  I loaded up your sub and drone and did a few docks and undocks.  I got the crash when the drone was approaching for a dock on the main sub.

The issue is caused by a NaN (not a number) result proliferating through the code.  Once the code has to rely on value to be an actual number, your game will crash.


Technical Info:
The crash is occurring in /Utils/Toolbox.cs:
"return Color.Lerp(gradient[(int)scaledT], gradient[(int)Math.Min(scaledT + 1, gradient.Length - 1)], (scaledT - (int)scaledT));"

In this case, scaledT = NaN.  I followed it back through the code as far back as this loop in Machines/Sonar.cs:

"foreach (AITarget t in **AITarget.List)..."

The issue here is that the SoundRange field of t is NaN (not a number).  This NaN just moves through the code, creating trouble, until it hits that line in Toolbox, where it tries to index into the NaN'th entry of an array.  Boom.

**Comment by mrjpjr** [comment]
Hey! Thanks for looking at it. I know the Devs have their hands full with getting those new and exciting changes rolled out as soon as possible.
Unfortunately I am not a programmer, so while I can follow what you are saying, the actual code portion is lost to me, as I don't have context for the variables listed (though I would assume Toolbox.cs is for the Toolbox and Sonar.cs is for the Transducer)
Are you able to explain the issue in game terms? For example: What item and/or function uses the string(s) you have highlighted? Is there anything I can change on the drone to prevent the crash from happening.

**Comment by jamiebidelia** [comment]
mrjpjr,

From what I know so far, the issue is caused by the sonar.  The sonar draws "blips" on the screen when you see things.  The "strength" of the thing you're seeing ends up being NaN (not a number), which is caused by the SoundRange field not being set properly in the code.  When I was looking at the debugger, the object that was causing the issue had the name "hull".

I would have to take a deeper look in the next day or so to see if I can follow the problem back a bit more.  I took a look at your ship in the sub editor, and I didn't see anything that caught my eye in terms of "change this."  Though I'm not the best at the sub editor myself.

**Comment by jamiebidelia** [comment]
mrjpjr,

So I think I have located the root cause of the crash you're having.  It's the engine on your drone, the RN 1.3.

Functional Side:
The Min Voltage of your Engine is 0.0.  This is causing a math error in the game.  If you change your Min Voltage to a non-zero number, it should be fine.  Most engines in the game seem to use 0.5, but you can go lower if you like.

![ShuttleEngine](https://user-images.githubusercontent.com/12755132/72214660-cde0e300-34d4-11ea-88a4-1ce6e86f982d.png)



Technical Side:
So the game does some math with voltages.  There is a field called MinVoltage, which the code calls "The minimum voltage required for the device to function."  On your drone, the Min Voltage you've given is 0.  Many engines in the game use a Min Voltage of 0.5; you can definitely go lower, but 0 is the issue.

in the shared code in engine.cs, ln 94, there is a force vector is computed including this term:
"Math.Min(Voltage / MinVoltage, 1.0f)"
Basically, it's saying "give me the minimum of the Voltage / MinVoltage, and 1."
In your case, Voltage and MinVoltage are 0.
0/0 = Not a Number (NaN).
It can't compare NaN and 1, so it returns NaN.
So the Min is returning NaN here, which blows up the code.

I'll try making a fix for this tomorrow morning and doing a test on it.  If I'm happy about it I'll ask the Barotrauma team to accept my change.

**Comment by mrjpjr** [comment]
Oh wow, that was amazing. Thanks! 

I had to set min voltage to 0 to overcome another issue of the drone depowering, even while controlling it, causing the drone to be non responsive to control commands. 

I'll have to try another workaround, but at least now I know the source of the issue. 

Thanks again!

**Comment by mrjpjr** [comment]
jamiebidelia,
It works!
I changed the min voltage of the engine and Transducer to >0 and over a period of 90 mins of tweaking and testing the game did not crash even once! There is still the issue of the drone de-powering, but I have made a workaround for it.
Now I can use my drone in game! Thank you so much!

**Comment by Regalis11** [comment]
I'm glad you managed to solve this! @jamiebidelia thank you so much for the debugging help. 

I thought I had taken this into account by clamping the force below zero: dividing a non-zero float value by zero results in Infinity, which gets clamped by Min as expected. What I didn't take into account was that 0.0f / 0.0f results in NaN, so if both the Voltage and MinVoltage were zero, it blew up the code. 

Should be fixed for the next patch now: https://github.com/Regalis11/Barotrauma-development/commit/a5c31ec00a04af3c04e207fc4c3fe7e4213e0b7f (private repo)
